pkg/controller/gcp_access_key_request.go

```diff
@@ -79,47 +54,120 @@ func (c *VaultController) runGCPAccessKeyRequestInjector(key string) error {
 		glog.Warningf("GCPAccessKeyRequest %s does not exist anymore", key)
 
 	} else {
-		gcpAccessReq := obj.(*api.GCPAccessKeyRequest).DeepCopy()
+		req := obj.(*api.GCPAccessKeyRequest).DeepCopy()
 
-		glog.Infof("Sync/Add/Update for GCPAccessKeyRequest %s/%s", gcpAccessReq.Namespace, gcpAccessReq.Name)
+		glog.Infof("Sync/Add/Update for GCPAccessKeyRequest %s/%s", req.Namespace, req.Name)
 
-		if gcpAccessReq.DeletionTimestamp != nil {
-			if core_util.HasFinalizer(gcpAccessReq.ObjectMeta, GCPAccessKeyRequestFinalizer) {
-				go c.runGCPAccessKeyRequestFinalizer(gcpAccessReq, finalizerTimeout, finalizerInterval)
+		if req.DeletionTimestamp != nil {
+			if core_util.HasFinalizer(req.ObjectMeta, KubeVaultFinalizer) {
+				return c.runGCPAccessKeyRequestFinalizer(req)
 			}
 		} else {
-			if !core_util.HasFinalizer(gcpAccessReq.ObjectMeta, GCPAccessKeyRequestFinalizer) {
+			if !core_util.HasFinalizer(req.ObjectMeta, KubeVaultFinalizer) {
 				// Add finalizer
-				_, _, err = patchutil.PatchGCPAccessKeyRequest(context.TODO(), c.extClient.EngineV1alpha1(), gcpAccessReq, func(binding *api.GCPAccessKeyRequest) *api.GCPAccessKeyRequest {
-					binding.ObjectMeta = core_util.AddFinalizer(binding.ObjectMeta, GCPAccessKeyRequestFinalizer)
-					return binding
+				_, _, err = patchutil.PatchGCPAccessKeyRequest(context.TODO(), c.extClient.EngineV1alpha1(), req, func(in *api.GCPAccessKeyRequest) *api.GCPAccessKeyRequest {
+					in.ObjectMeta = core_util.AddFinalizer(in.ObjectMeta, KubeVaultFinalizer)
+					return in
 				}, metav1.PatchOptions{})
 				if err != nil {
-					return errors.Wrapf(err, "failed to set GCPAccessKeyRequest finalizer for %s/%s", gcpAccessReq.Namespace, gcpAccessReq.Name)
+					return errors.Wrapf(err, "failed to add GCPAccessKeyRequest finalizer for %s/%s", req.Namespace, req.Name)
 				}
 			}
 
 			var condType string
-			for _, c := range gcpAccessReq.Status.Conditions {
-				if c.Type == kmapi.ConditionRequestApproved || c.Type == kmapi.ConditionRequestDenied {
+			for _, c := range req.Status.Conditions {
+				if (c.Type == kmapi.ConditionRequestApproved && c.Status == kmapi.ConditionTrue) || (c.Type == kmapi.ConditionRequestDenied && c.Status == kmapi.ConditionTrue) {
 					condType = c.Type
 				}
 			}
 
+			// If condition type is not set yet, set the phase to "WaitingForApproval".
+			if condType == "" {
+				_, err := patchutil.UpdateGCPAccessKeyRequestStatus(
+					context.TODO(),
+					c.extClient.EngineV1alpha1(),
+					req.ObjectMeta,
+					func(status *api.GCPAccessKeyRequestStatus) *api.GCPAccessKeyRequestStatus {
+						if status.Phase == "" {
+							status.Phase = api.RequestStatusPhaseWaitingForApproval
+						}
+						status.ObservedGeneration = req.Generation
```

Comment:
Don't set ObservedGeneration here because `status` could be from a newer object.